### PR TITLE
[TBS] Wait for ProcessBatch before stopping sampler

### PR DESCRIPTION
**Pull request by stuartnelson3**
## Motivation/summary

prevent data loss from stopping the tail-based sampler while batches are still being processed

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

## How to test these changes

I wasn't able to reproduce the issue as described in https://github.com/elastic/apm-server/issues/8339, ie. while sending a continuous stream of transactions and then disabling TBS, I never saw `processor is stopped` in the logs.

With the change I added, I did see the log message indicating that a batch was received in the TBS while it was stopping, `received batch while stopping; waiting`; after the configured timeout the "no batches received" message was logged and the TBS stopped.

1. start the stack
    ```
    ~/w/apm-integration-testing (main=)
    # ./scripts/compose.py start 8.4.0 --remove-orphans --no-apm-server
    ```
2. start apm-server with tbs enabled
3. send a stream of transactions to the apm-server. I used the following go program:

```go
package main

import (
	"context"
	"fmt"
	"os"

	"go.elastic.co/apm/v2"
)

func main() {
	version := "undefined"
	if len(os.Args) > 1 {
		version = os.Args[1]
	}
	ctx := context.Background()
	name := fmt.Sprintf("apm-server-%s", version)
	tracer := apm.DefaultTracer()
	for {
		tx := tracer.StartTransaction("tx1", "type")
		ctx = apm.ContextWithTransaction(ctx, tx)
		span, ctx := apm.StartSpan(ctx, "span1", "type")
		span.Context.SetLabel("enabled", true)
		span.Context.SetLabel("string", "string")
		span.Context.SetLabel("string2", "string2")
		span.Context.SetLabel("string3", "string3")
		span.Context.SetLabel("numbers", 1234.56)
		span.End()
		tx.End()

		tx = tracer.StartTransaction("tx2", "type")
		ctx = apm.ContextWithTransaction(ctx, tx)
		span, ctx = apm.StartSpan(ctx, "span2", "type")
		span.Context.SetLabel("other", "label")
		span.End()
		tx.End()

		tracer.Flush(nil)
	}
	fmt.Printf("%s: %+v\n", name, tracer.Stats())
}
```
4. while still streaming transactions, disable tbs and reload the apm-server
5. check the log output for something like this:
```
{"log.level":"info","@timestamp":"2022-06-15T14:44:31.059+0200","log.logger":"sampling","log.origin":{"file.name":"sampling/processor.go","file.line":328},"message":"received batch while stopping; waiting","service.name":"apm-server","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2022-06-15T14:44:36.059+0200","log.logger":"sampling","log.origin":{"file.name":"sampling/processor.go","file.line":335},"message":"no batches received; stopping","service.name":"apm-server","ecs.version":"1.6.0"}
```

Note: I ran this locally using https://github.com/elastic/apm-server/commit/fb74c6e3048ebf33e1a6a8e8f0fad767d6e836c2, so my actual flow was to update `apm-server.dev.yml` and `kill -s USR1 $(pidof apm-server)`, but the update via managed apm-server + apm-integration should be the same.

## Related issues

closes https://github.com/elastic/apm-server/issues/8339

### CODE REVIEW

**Comment by stuartnelson3** [comment]
Couple comments/questions:
- I'd love for some way to indicate to the batch processor that no more batches will come through. If you could propagate some sort of "we're done" when [the server is stopped](https://github.com/elastic/apm-server/blob/main/beater/server.go#L242-L247), for example, then the whole 5sec timer I've added here could be dropped in favor of something more deterministic. I'm going to look into this, but I leave it up to the reviewer to say if this is something I should continue pursuing or not (not sure what this would imply for larger code changes, it might be better for a different PR)
- Any suggestion on how to write a test for this would be much appreciated. Based on the current implementation ("let's wait for a bit until we don't get any more messages"), my idea would be depending on the non-determinism of "call p.Stop(), wait long enough for p.Stop() to progress but not so long that the timer has fired, process a batch, .."

**Comment by stuartnelson3** [comment]
here's an idea for using a `ctx.Value` to indicate no more batches will be coming, if this is something to pursue let me know:
```diff
diff --git a/beater/server.go b/beater/server.go
index 125ee9e54..97dc68e06 100644
--- a/beater/server.go
+++ b/beater/server.go
@@ -114,6 +114,7 @@ type server struct {
 	logger                *logp.Logger
 	cfg                   *config.Config
 	agentcfgFetchReporter agentcfg.Reporter
+	batchProcessor        model.BatchProcessor
 
 	httpServer *httpServer
 	grpcServer *grpc.Server
@@ -181,6 +182,7 @@ func newServer(args ServerParams, listener net.Listener) (server, error) {
 		httpServer:            httpServer,
 		grpcServer:            grpcServer,
 		agentcfgFetchReporter: agentcfgFetchReporter,
+		batchProcessor:        batchProcessor,
 	}, nil
 }
 
@@ -229,6 +231,9 @@ func newGRPCServer(
 	return srv, nil
 }
 
+// ProcessingStopped indicates that no more batches will be sent to the batch processor.
+type ProcessingStopped struct{}
+
 func (s server) run(ctx context.Context) error {
 	s.logger.Infof("Starting apm-server [%s built %s]. Hit CTRL-C to stop it.", version.Commit(), version.BuildTime())
 	defer s.logger.Infof("Server stopped")
@@ -243,6 +248,11 @@ func (s server) run(ctx context.Context) error {
 		<-ctx.Done()
 		s.grpcServer.GracefulStop()
 		s.httpServer.stop()
+		ctx = context.WithValue(context.Background(), ProcessingStopped{}, struct{}{})
+		if err := s.batchProcessor.ProcessBatch(ctx, new(model.Batch)); err != nil {
+			s.logger.Errorf("error sending closing batch: %+v", err)
+			return err
+		}
 		return nil
 	})
 	if err := g.Wait(); err != http.ErrServerClosed {
diff --git a/x-pack/apm-server/sampling/processor.go b/x-pack/apm-server/sampling/processor.go
index 6a09074a0..1a6213a12 100644
--- a/x-pack/apm-server/sampling/processor.go
+++ b/x-pack/apm-server/sampling/processor.go
@@ -17,6 +17,7 @@ import (
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
 
+	"github.com/elastic/apm-server/beater"
 	logs "github.com/elastic/apm-server/log"
 	"github.com/elastic/apm-server/model"
 	"github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage"
@@ -142,11 +143,18 @@ func (p *Processor) ProcessBatch(ctx context.Context, batch *model.Batch) error
 	if p.storage == nil {
 		return ErrStopped
 	}
+	// Check for the sentinel value indicating we won't receive any more
+	// batches.
+	if v := ctx.Value(beater.ProcessingStopped{}); v != nil {
+		p.logger.Info("processing stopped signal received")
+		select {
+		case p.processing <- struct{}{}:
+		default:
+			p.logger.Error("cannot send on processing channel")
+		}
+		return nil
+	}
 	events := *batch
-	go func() {
-		// indicate to p.Stop() that we're still processing batches
-		p.processing <- struct{}{}
-	}()
 	for i := 0; i < len(events); i++ {
 		event := &events[i]
 		var report, stored bool
@@ -179,10 +187,6 @@ func (p *Processor) ProcessBatch(ctx context.Context, batch *model.Batch) error
 		p.updateProcessorMetrics(report, stored)
 	}
 	*batch = events
-	select {
-	case <-p.processing:
-	default:
-	}
 	return nil
 }
 
@@ -325,15 +329,12 @@ func (p *Processor) Stop(ctx context.Context) error {
 		for {
 			select {
 			case <-p.processing:
-				p.logger.Info("received batch while stopping; waiting")
+				p.logger.Info("processing complete")
 				if !t.Stop() {
 					<-t.C
 				}
-				t.Reset(d)
-				continue
 			case <-t.C:
-				p.logger.Info("no batches received; stopping")
-				return
+				p.logger.Infof("processing not stopped after %s; continuing", d)
 			}
 		}
 	}()
@@ -351,7 +352,7 @@ func (p *Processor) Stop(ctx context.Context) error {
 	p.storage = nil
 	// Guarded by p.storageMu; subsequent calls to p.Stop() or
 	// p.ProcesseBatch() will early return before attempting to interact
-	// with p.process.
+	// with p.processing
 	close(p.processing)
 	return nil
 }
```

**Comment by axw** [comment]
> I'd love for some way to indicate to the batch processor that no more batches will come through.

Could we have this be based off the shutdown of the HTTP and gRPC servers?

At the moment, we use the same context for both the servers and for the processors: https://github.com/elastic/apm-server/blob/2a7b40f83173891532441ea55b39b6fadfa66f4c/x-pack/apm-server/main.go#L180-L207 When the [context is cancelled](https://github.com/elastic/apm-server/blob/2a7b40f83173891532441ea55b39b6fadfa66f4c/beater/beater.go#L350), that will propagate to the servers as well as the processors, in parallel.

**Comment by stuartnelson3** [comment]
@axw do you have an example program you were using to trigger the "processor is stopped" message? I've tried adding multiple go routines to the sample I have in the description and still haven't been able to trigger it

**Comment by axw** [comment]
@stuartnelson3 recently I've been generating load with this little program. I think it's what I used to trigger the message, but my memory is fuzzy.

```go
package main

import (
        "time"

        "go.elastic.co/apm/v2"
)

func main() {
        tracer := apm.DefaultTracer()
        defer tracer.Flush(nil)
        for {
                for i := 0; i < 500; i++ {
                        once(tracer)
                }
                tracer.Flush(nil)
                time.Sleep(time.Millisecond * 100)
        }
}

func once(tracer *apm.Tracer) {
        tx := tracer.StartTransaction("name", "type")
        defer tx.End()

        span := tx.StartSpan("name", "type", nil)
        defer span.End()

        time.Sleep(10 * time.Millisecond)

        span.Context.SetDestinationService(apm.DestinationServiceSpanContext{
                Resource: "dest_resource",
        })
}
```

File: x-pack/apm-server/main.go [hunk 1]
```diff
@@ -178,6 +178,8 @@ func runServerWithProcessors(ctx context.Context, runServer beater.RunServerFunc
 	runServer = beater.WrapRunServerWithProcessors(runServer, batchProcessors...)
 
 	g, ctx := errgroup.WithContext(ctx)
+	serverctx, cancel := context.WithCancel(context.Background())
+	ctx = context.WithValue(ctx, beater.ServerStopped{}, cancel)
```
Comment (axw) [inline x-pack/apm-server/main.go:182]:
```suggestion
	serverStopped := make(chan struct{})
```

In the processor.Stop goroutine:

```go
		g.Go(func() error {
			<-serverStopped
			stopctx := context.Background()
```

In the runServer goroutine:

```go
	g.Go(func() error {
		defer close(serverStopped)
		return runServer(ctx, args)
	})
```

Comment (stuartnelson3) [reply x-pack/apm-server/main.go]: that's much better

**Review by axw — APPROVED** [review]
Thanks!

**Comment by lahsivjar** [comment]
Confirmed successfully on 8.4.0 BC ([45083f33](https://github.com/elastic/apm-server/commits/45083f332df130339dd2174f235409ac5f693469)) by using the following steps:

1. Send a continuous stream of events to APM-Server with TBS enabled
2. While sending the stream of events, disable TBS
3. Observe the log for the order of events, the server should be stopped first followed by all processors.
    <details>
         <summary>Snapshot of observation</summary>
         
         order_of_events_on_reload

    </details>

**Pull request merged**